pkg/openid4ci/openid4ci.go

```diff
@@ -73,17 +76,181 @@ func (i *Interaction) Authorize() (*AuthorizeResult, error) {
+func (i *Interaction) RequestCredential(credentialRequestOpts *CredentialRequestOpts) ([]CredentialResponse, error) {
+	if i.initiationRequest.UserPINRequired && credentialRequestOpts.UserPIN == "" {
+		return nil, errors.New("issuance request indicates that a user PIN is required, " +
+			"but no user PIN was provided")
+	}
+
+	metadata, err := i.getIssuerMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get issuer metadata: %w", err)
+	}
+
+	tokenEndpointURL, err := i.buildTokenURLWithQueryParameters(credentialRequestOpts.UserPIN, metadata.TokenEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build token URL with query parameters: %w", err)
+	}
+
+	tokenResp, err := i.getTokenResponse(tokenEndpointURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token response: %w", err)
+	}
+
+	credentialResponses := make([]CredentialResponse, len(i.initiationRequest.CredentialTypes))
+
+	for index, credentialType := range i.initiationRequest.CredentialTypes {
+		credentialResponse, err := i.getCredentialResponse(credentialType, metadata.CredentialEndpoint, tokenResp.AccessToken)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get credential response: %w", err)
+		}
+
+		credentialResponses[index] = *credentialResponse
+	}
+
+	return credentialResponses, nil
+}
+
+func (i *Interaction) getIssuerMetadata() (*issuerMetadata, error) {
+	metadataEndpoint := i.initiationRequest.IssuerURI + "/.well-known/openid-configuration"
+
+	// TODO: Implement trusted list type of mechanism? The gosec warning (correctly) warns about using a variable URL.
+	response, err := http.Get(metadataEndpoint) //nolint: noctx,gosec // TODO: To be re-evaluated later
+	if err != nil {
+		return nil, err
+	}
+
+	responseBytes, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received status code [%d] with body [%s] from issuer's "+
+			"OpenID configuration endpoint", response.StatusCode, string(responseBytes))
+	}
+
+	defer func() {
+		errClose := response.Body.Close()
+		if errClose != nil {
+			println(fmt.Sprintf("failed to close response body: %s", errClose.Error()))
+		}
+	}()
+
+	var metadata issuerMetadata
+
+	err = json.Unmarshal(responseBytes, &metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response from the issuer's "+
+			"OpenID configuration endpoint: %w", err)
+	}
+
+	return &metadata, err
+}
+
+func (i *Interaction) buildTokenURLWithQueryParameters(userPIN, tokenEndpoint string) (*url.URL, error) {
+	tokenEndpointURL, err := url.Parse(tokenEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	query := tokenEndpointURL.Query()
+
+	query.Add("grant_type", "urn:ietf:params:oauth:grant-type:pre-authorized_code")
+	query.Add("pre-authorized_code", i.initiationRequest.PreAuthorizedCode)
+	query.Add("user_pin", userPIN)
+
+	tokenEndpointURL.RawQuery = query.Encode()
+
+	return tokenEndpointURL, nil
+}
```

Comment:
this is POST request, why are we adding user secrets in URL

Reply:
Fixed - moved it to the body